spec/requests/customers_spec.rb

```diff
@@ -0,0 +1,92 @@
+      FactoryBot.create_list(:customer, 10)
+      get customers_path
+      expect(response.status).to eq(200)
+    end
+  end
+
+  describe "get customer_path" do
+    it "renders the :show template" do
+      customer = FactoryBot.create(:customer)
+      get customer_path(id: customer.id)
+      expect(response.status).to eq(200)
+    end
+
+    it "redirects to the index path if the customer id is invalid" do
+      get customer_path(id: 5000)  # An ID that doesn't exist.
+      expect(response).to redirect_to customers_path
+    end
+  end
+
+  describe "get new_customer_path" do
+    it "renders the :new template" do
+      get new_customer_path
+      expect(response.status).to eq(200)
+    end
+  end
+
+  describe "get edit_customer_path" do
+    it "renders the :edit template" do
+      customer = FactoryBot.create(:customer)
+      get edit_customer_path(id: customer.id)
+      expect(response.status).to eq(200)
+    end
+  end
+
+  describe "post customers_path with valid data" do
+    it "saves a new entry and redirects to the show path for the entry" do
+      customer_attributes = FactoryBot.attributes_for(:customer)
+      expect { post customers_path, params: {customer: customer_attributes}
+        }.to change(Customer, :count)
+      expect(response).to redirect_to customer_path(id: Customer.last.id)
+    end
+  end
+
+  describe "post customers_path with invalid data" do
+    it "does not save a new entry or redirect" do
+      customer_attributes = FactoryBot.attributes_for(:customer)
+      customer_attributes.delete(:first_name)
+      expect { post customers_path, params: {customer: customer_attributes}
+        }.to_not change(Customer, :count)
+      expect(response.status).to eq(200)
+    end
+  end
+
+  describe "put customer_path with valid data" do
+    it "updates an entry and redirects to the show path for the customer" do
+      customer = FactoryBot.create(:customer)
+      # TODO Is this right? I feel like I'm mixing references to `customer`
+      #  and `:customer` in an odd way.
+      customer_attributes = FactoryBot.attributes_for(:customer)
+      customer_attributes[:first_name] = "Updated!"
+      expect { put customer_path(id: customer.id), params: {customer: customer_attributes}
+    }.to change {Customer.last.first_name}
+      expect(response).to redirect_to customer_path(id: customer.id)
+    end
+  end
+
+  describe "put customer_path with invalid data" do
+    it "does not update the customer record or redirect" do
+      customer = FactoryBot.create(:customer)
+      # TODO Is this right? I feel like I'm mixing references to `customer`
+      #  and `:customer` in an odd way.
+      customer_attributes = FactoryBot.attributes_for(:customer)
+      customer_attributes[:phone] = "invalid-phone"
+      expect { put customer_path(id: customer.id), params: {customer: customer_attributes}
+    }.to_not change {Customer.last.phone}
+      expect(response.status).to eq(200)
+    end
```

Comment:
Do these look right? It feels like I'm mixing references to `customer` and `:customer`. I don't quite understand how `:customer` is used in these tests.

Reply:
Jamie, these are correct.  ':customer' is here being used as a label for FactoryBot, so that it knows which factory to invoke.  'customer' is the object you actually create.  You can use the customer object instead of Customer.last, but if you do so, you have to do customer.reload to get the current state (the state after the change).  Either way is correct.

Reply:
Ah, that makes sense! Thanks so much for checking this out and helping me learn!